### DISCUSSION TITLE
Issue with Right Shift + D Disconnecting in FreeRDP

### DISCUSSION THREAD

**Discussion by kaiserfr** [FreeRDP/FreeRDP#10588] [Q&A] [answered]
Hi everyone,

I'm facing an issue where every time I type a capital "D" using the right Shift key, the connection closes. According to the command-line help, this is because Right Shift + D is the default shortcut to disconnect.

This is causing me constant frustration since I accidentally disconnect whenever I need to type a capital "D."

From what I’ve read, I can create an sdl-freerdp.json file to modify the shortcuts. Could someone please provide an example of how this .json file should look to change or disable this specific shortcut?

Thanks in advance!

**Comment by akallabeth** [comment, marked as answer]
1. `sdl-freerdp` help output as well as manpage document the location and options
2. This simple `JSON` switches the `KMOD_RSHIFT` to `KMOD_RCTRL`: `{ "SDL_KeyModMask": ["KMOD_RCTRL"] }`

**Reply by kaiserfr** [reply]
I created an sdl-freerdp.json on the path the commandline-help provided me with: 
`/home/<username>/.var/app/com.freerdp.FreeRDP/config/freerdp/sdl-freerdp.json`

The content of the JSON file is `{ "SDL_Disconnect": ["SDL_SCANCODE_RETURN2"] }`

I would expect, that  the `<right shift> + D` shortcut is now deactivated (replaced by `right shift + return2` and I am not disconnected anymore. But that's not the case. What may have gone wrong?

**Reply by akallabeth** [reply]
you try to set an *array* where a *string* is expected.
try `{ "SDL_Disconnect": "SDL_SCANCODE_RETURN2" }`
hint: `"SDL_KeyModMask"` is, as the name suggests, an array of keys (like `<ctrl>+<alt>` combo used by `Qemu/KVM`) whereas the others are simple assignments to a single key

**Reply by kaiserfr** [reply]
Now it works - thanks a lot !

**Comment by PatrickGit7** [comment]
This solution seems to have been broken in the latest release. I had previously had success with adding `{ "SDL_KeyModMask": ["KMOD_RCTRL"] }` to sdl-freerdp.json. However, upon updating to version 3.19, `<right shift> + D` is back to disconnecting my RDP session (and I'm also now noticing `<right shift> + R` and `<right shift> + G` also send other commands I don't want.

I also tried adding `{ "SDL_Disconnect": "SDL_SCANCODE_RETURN2" }` to the same .json, with no luck.

Does anyone have any further guidance here?

It really is disruptive to have commands sent via quite common normal typing keystrokes (e.g. simple capital punctuation) - maybe this behavior could be updated for future releases?

**Reply by akallabeth** [reply]
@PatrickGit7 https://github.com/FreeRDP/FreeRDP/issues/12051

**Comment by t00mietum** [comment]
I can't seem to stop freerdp from killing the session when I press CTRL+ALT+D, which is a hotkey in an application I use everywhere.

I tried following the suggestions in various issues, and in this comment thread. In order (all in `~/.config/freerdp/sdl-freerdp.json`):

1. `{ "SDL_KeyModMask": [ "SDL_KMOD_NONE" ] }`
2. `{ "SDL_Disconnect": "SDL_SCANCODE_RETURN2" }`
3. Both: `{ "SDL_KeyModMask": [ "SDL_KMOD_NONE" ], "SDL_Disconnect": "SDL_SCANCODE_RETURN2" }`

None work. That key combination still kills the client.

I don't need _any_ RDP hotkeys. I never run in fullscreen.

Due to issues with xrdp on the remote host that also seem to trouble/baffle users, when the session is killed in this way, it cannot be resumed. Not even restarting xrdp and/or xrdp-sesman services do it, the machine has to be rebooted. (Presumably restarting lightdm might also do the trick, but either way is massively disruptive to anyone using the machine remotely.)

Closing the remote window from the outside seems to always result in being able to reconnect to the same session later.

But when killed via the hotkey from within, it somehow hoses the server up.

IOW, using a common hotkey I use all the time, kills not just my session but effectively screws the server up.

That these hotkeys are (or were) hard-coded is a bit baffling. I wish I remembered C++!

**Edit**: Restarting lightdm (my display manager), xrdp, and xrdp-sesman don't resolve the remote problem after the freerdp kills itself with CTRL+ALT+D. Kill -9 on anything resembling "rdp" doesn't do it. A full reboot of the remote host is required. Arg.

**Reply by thecooldaniel** [reply]
I was able to disable all SDL scancodes:
`{ "SDL_KeyModMask": "KMOD_NONE" }`

It seems that `sdl-freerdp` is expecting the name of the `SDL_Keymod` without the leading `SDL_` as they are defined here: https://wiki.libsdl.org/SDL3/SDL_Keymod 

I am using version 3.26.0

**Reply by akallabeth** [reply]
@thecooldaniel no, both are valid (assuming you are not using an old version, >= 3.17.2 has that)

**Reply by thecooldaniel** [reply]
@akallabeth I'm not sure what to say, "SDL_KMOD_NONE" did not work, but "KMOD_NONE" did. 🤷
edit: I am on Fedora 43 and using the version of `freerdp` available via `dnf`.

**Reply by akallabeth** [reply]
@thecooldaniel using it on fedora 44 and debian 13 for my config.

**Reply by t00mietum** [reply]
So in my case it was probably that it turns out I was using the flatpak version of freerdp. I kind of hid this fact from myself by using a little bash wrapper named `xfreerdp`.

Because running flatpak CLI program is no fun.

And why flatpak? Because I used to run Debian Testing, and at one point freerdp got yanked, as random packages are now and then. Flatpak is the way to go on Testing. Except for CLI programs.

Anyway, the config file that didn't work for the secret flatpak program because the flatpak version couldn't even see the directory, now works with the official sdl-freerdp version. Say goodbye to those pesky conflicting hotkeys!

**Reply by akallabeth** [reply]
@t00mietum then you need to check if your config directory was correct.
```
flatpak run com.freerdp.FreeRDP /? |grep Location
  The location is a system configuration file. Location for current machine is /app/etc/FreeRDP/sdl-freerdp.json
  The location is a per user file. Location for current user is /home/username/.var/app/com.freerdp.FreeRDP/config/freerdp/sdl-freerdp.json
```

**Comment by acidkewpie** [comment]
I was finding the docs on SDL hotkeys extremely elusive, no example files anywhere I could see. But from this issue I have this configuration which serves as an example sdl-freerdp.json file for any other searchers:

```
{
   "SDL_KeyModMask": ["KMOD_LGUI"],
   "SDL_Fullscreen": "SDL_SCANCODE_F",
   "SDL_Disconnect": "SDL_SCANCODE_Q"
}

```

Where `LGUI` is the windows / mod / super key.
